src/mozilla_sec_eia/ex_21/inference.py

```diff
@@ -0,0 +1,342 @@
+        predictions.append(pred)
+        all_output_df = pd.concat([all_output_df, output_df])
+    all_output_df = all_output_df[["id", "subsidiary", "loc", "own_per"]]
+    return logits, predictions, all_output_df
+
+
+class LayoutLMInferencePipeline(Pipeline):
+    """Pipeline for performing inference with fine-tuned LayoutLM."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize LayoutLMInferencePipeline."""
+        super().__init__(*args, **kwargs)
+
+    def _sanitize_parameters(self, **kwargs):
+        preprocess_kwargs = {}
+        if "maybe_arg" in kwargs:
+            preprocess_kwargs["maybe_arg"] = kwargs["maybe_arg"]
+        return preprocess_kwargs, {}, {}
+
+    def preprocess(self, example):
+        """Encode and tokenize model inputs."""
+        image = example["image"]
+        words = example["tokens"]
+        boxes = example["bboxes"]
+        encoding = self.tokenizer(
+            image,
+            words,
+            boxes=boxes,
+            return_tensors="pt",
+            truncation=True,
+            padding="max_length",
+            max_length=512,  # this is the maximum max_length
+            stride=128,
+            return_offsets_mapping=True,
+            return_overflowing_tokens=True,
+        )
+        model_inputs = {}
+        model_inputs["raw_encoding"] = encoding.copy()
+        model_inputs["example"] = example
+        model_inputs["offset_mapping"] = encoding.pop("offset_mapping")
+        model_inputs["sample_mapping"] = encoding.pop("overflow_to_sample_mapping")
+        # TODO: do we actually need to make these into ints?
+        encoding["input_ids"] = encoding["input_ids"].to(torch.int64)
+        encoding["attention_mask"] = encoding["attention_mask"].to(torch.int64)
+        encoding["bbox"] = encoding["bbox"].to(torch.int64)
+        encoding["pixel_values"] = torch.stack(encoding["pixel_values"])
+        model_inputs["encoding"] = encoding
+        return model_inputs
+
+    def _forward(self, model_inputs):
+        encoding = model_inputs["encoding"]
+        if torch.cuda.is_available():
+            encoding.to("cuda")
+            self.model.to("cuda")
+        # since we're doing inference, we don't need gradient computation
+        with torch.no_grad():
+            output = self.model(**encoding)
+            return {
+                "logits": output.logits,
+                "predictions": output.logits.argmax(-1).squeeze().tolist(),
+                "raw_encoding": model_inputs["raw_encoding"],
+                "example": model_inputs["example"],
+            }
+
+    def postprocess(self, all_outputs):
+        """Return logits, model predictions, and the extracted dataframe."""
+        logits = all_outputs["logits"]
+        predictions = all_outputs["logits"].argmax(-1).squeeze().tolist()
+        output_df = self.extract_table(all_outputs)
+        return logits, predictions, output_df
+
+    def extract_table(self, all_outputs):
```

Comment:
This first implementation of this function is super basic. It essentially organizes bounding boxes from left to right and top to bottom. Then every time a new subsidiary is encountered, it assigns a new group or "row" in the extracted dataframe. Next, location and ownership percentage words/labeled entities in between these subsidiary groups are assigned to a row/group. Finally, this is all formatted into a dataframe.

Reply:
Seems like a reasonable place to start. I'm excited to get validation metrics so we can actually work on refining the model.